Add information to path planning re: Bezier curves (#680)
PX4/px4_user_guide: en/computer_vision/path_planning_interface.md

> # Path Planning Offboard Interface

Author: So I'm not sure if the Bezier curve works under the offboard interface. It might, but I'd be pleasantly surprised.

Reviewer: OK, now I understand. This title was supposed to indicate that Path Planning is an interface driven by a companion computer, not that it is an interface that is active in offboard mode. I will just rename it Path Planning Interface to reduce this confusion.

Thanks

- There are two [MAVLink Path Planning Protocol](https://mavlink.io/en/services/trajectory.html) interfaces:
  > **Warning** Route planning software should not mix these interfaces while executing a task (PX4 will use the last received message of either type).

  - The [Waypoint Trajectory Interface](#waypoint_interface) uses [TRAJECTORY_REPRESENTATION_WAYPOINTS](https://mavlink.io/en/messages/common.html#TRAJECTORY_REPRESENTATION_WAYPOINTS) to sent the current and next waypoint, and to receive a stream of setpoints for the new path.
  - The [Bezier Trajectory Interface](#bezier_interface) uses [TRAJECTORY_REPRESENTATION_BEZIER](https://mavlink.io/en/messages/common.html#TRAJECTORY_REPRESENTATION_BEZIER) to specify the moving position setpoint of the vehicle over a given amount of time.
- The [HEARTBEAT/Connection Protocol](https://mavlink.io/en/services/heartbeat.html) is used for "proof of life" detection.
- [LOCAL_POSITION_NED](https://mavlink.io/en/messages/common.html#LOCAL_POSITION_NED) and [ALTITUDE](https://mavlink.io/en/messages/common.html#ALTITUDE) send the vehicle local position and altitude, respectively.

> Path planning is enabled on PX4 in automatic modes (landing, takeoff, hold, mission, return) if [COM_OBS_AVOID=1](../advanced_config/parameter_reference.md#COM_OBS_AVOID).

Reviewer: > Whoops, the Bezier isn't available in Offboard (I think). That would need to be clarified as well. It is only available in COM_OBS_AVOID=1 mode.

@jkflying I added this paragraph to cover the case - ie that it is enabled in this set of modes when COM_OBS_AVOD is set to one, and that the planner must supply setpoints even if it cannot do proper planning for the mode. 

OK?

Author: LGTM.

  This means that offboard features that use different planners cannot be enabled on the same vehicle. a vehicle at the same time (e.g. a vehicle can support obstacle avoidance and collision prevent, but not also safe landing - or visa versa).

## Trajectory Interface {#waypoint_interface)

> PX4 sends information about the *desired path* to the companion computer (when `COM_OBS_AVOID=1`, in modes for which the path planning interface has been integrated), and receives back a stream of setpoints for the *planned path* from the path planning software.

Author: ```suggestion
PX4 sends information about the *desired path* to the companion computer (when `COM_OBS_AVOID=1`, in _auto_ modes), and receives back a stream of setpoints for the *planned path* from the path planning software.
```

## Bezier Trajectory Interface {#bezier_interface)

If the obstacle avoidance interface is enabled, it can also receive mavlink messages in the [TRAJECTORY_REPRESENTATION_BEZIER](https://mavlink.io/en/messages/common.html#TRAJECTORY_REPRESENTATION_BEZIER) format. This is parsed as follows:

- The number of Bezier control points determines the degree of the Bezier curve. For example, 3 points makes a quadratic bezier curve with constant acceleration.
- The Bezier curve must be the same degree in x,y,z and yaw, with all Bezier control points finite
- The `delta` array should have the value corresponding with the last Bezier control point indicate the duration that the waypoint takes to execute the curve to that point, from beginning to end. Other values in the `delta` array are ignored.
- The timestamp of the mavlink message should be the time that the curve starts, and communication delay and clock mismatch will be compensated for on the flight controller via the timesync mechanism.

> - The control points should all be specified in local coordinates.

Reviewer: This means FRD relative to vehicle center-of-gravity and as shown here right?: https://docs.px4.io/master/en/getting_started/px4_basic_concepts.html#heading-and-directions ?

Author: No, it is NED. This is the same frame as the waypoint trajectory message, BTW.

> - Bezier curves do not only expire after 0.5s like the waypoint interface, but also after the execution time of the bezier curve has been reached. Be careful not to specify such short times for the bezier curve that you cannot send the next one in time.

Reviewer: 1. What happens if the curve expires? Is this the same as message not being received for 0.5s? 
2. Are the fault handling rules for losing the companion-side otherwise the same [if so, I might move the section "Companion Failure Handing" down to a common area.

If the obstacle avoidance interface is enabled, it can also receive MAVLink messages in the [TRAJECTORY_REPRESENTATION_BEZIER](https://mavlink.io/en/messages/common.html#TRAJECTORY_REPRESENTATION_BEZIER) format.
This is parsed as follows:

- The number of bezier control points determines the degree of the bezier curve.
  For example, 3 points makes a quadratic bezier curve with constant acceleration.
- The bezier curve must be the same degree in x, y, z, and yaw, with all bezier control points finite
- The `delta` array should have the value corresponding with the last bezier control point indicate the duration that the waypoint takes to execute the curve to that point, from beginning to end.
  Other values in the `delta` array are ignored.
- The timestamp of the MAVLink message should be the time that the curve starts, and communication delay and clock mismatch will be compensated for on the flight controller via the timesync mechanism.
- The control points should all be specified in local coordinates.
- Bezier curves do not only expire after 0.5s like the waypoint interface, but also after the execution time of the bezier curve has been reached.

>   Be careful not to specify such short times for the bezier curve that you cannot send the next one in time.

Reviewer: 1. What happens if the curve expires? Is this the same as message not being received for 0.5s? 
2. Are the fault handling rules for losing the companion-side otherwise the same [if so, I might move the section "Companion Failure Handing" down to a common area.

Author: 1. Yes

2. Yes. Go for it :+1:

  For example, 3 points makes a quadratic bezier curve with constant acceleration.
- The bezier curve must be the same degree in x, y, z, and yaw, with all bezier control points finite
- The `delta` array should have the value corresponding with the last bezier control point indicate the duration that the waypoint takes to execute the curve to that point, from beginning to end.
  Other values in the `delta` array are ignored.
- The timestamp of the MAVLink message should be the time that the curve starts, and communication delay and clock mismatch will be compensated for on the flight controller via the timesync mechanism.
- The control points should all be specified in local coordinates.
- Bezier curves do not only expire after 0.5s like the waypoint interface, but also after the execution time of the bezier curve has been reached.
  Be careful not to specify such short times for the bezier curve that you cannot send the next one in time.

## Supported Hardware

Tested companion computers and cameras are listed in [PX4/avoidance](https://github.com/PX4/avoidance#run-on-hardware).

<!-- ## Further Information -->

> <!-- @mrivi and @jkflying are the experts! -->

Author: :scream:
